### PR TITLE
[th2-3147] timestamps inside batch and between batches should not decrease

### PR DESCRIPTION
**Pull request by andrey-drobynin**

### CODE REVIEW

File: src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java [hunk 1]
```diff
@@ -183,15 +187,34 @@ public final void handle(T messageBatch) {
...
+
+            long firstSequence = extractSequence(firstMessage);
             long lastSequence = extractSequence(lastMessage);
-            SessionKey sessionKey = createSessionKey(lastMessage);
-            SessionData sessionData = sessionToHolder.computeIfAbsent(sessionKey, ignore ->
-                    new SessionData(cradleStorage.getObjectsFactory()::createMessageBatch));
-            long prevLastSeq = sessionData.getAndUpdateSequence(lastSequence);
-            if (prevLastSeq >= firstSequence) {
-                logger.error("Duplicated batch found: {}", shortDebugString(messageBatch));
+            long previousBatchLastSequence = sessionData.getAndUpdateLastSequence(lastSequence);
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java:199]: We must update sequence number and timestamps atomically

File: src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java [hunk 2]
```diff
@@ -183,15 +187,34 @@ public final void handle(T messageBatch) {
...
+            if (previousBatchLastSequence >= firstSequence) {
+                logger.error(
+                        "Found batch with less or equal sequence: {}. Last sequence from previous batch: {}",
+                        shortDebugString(messageBatch),
+                        previousBatchLastSequence
+                );
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java:205]: Please check isErrorEnabled before this operation

File: src/main/java/com/exactpro/th2/mstore/MessageBatchStore.java [hunk 3]
```diff
@@ -65,6 +67,11 @@ protected SessionKey createSessionKey(Message message) {
         return new SessionKey(messageID.getConnectionId().getSessionAlias(), toCradleDirection(messageID.getDirection()));
     }
 
+    @Override
+    protected Instant extractTimestamp(Message message) {
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/mstore/MessageBatchStore.java:71]: Maybe write Timestamp comparator to avoid convertation

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

File: src/test/java/com/exactpro/th2/mstore/TestCaseMessageStore.java [hunk 4]
```diff
@@ -125,18 +129,13 @@ protected MessageID createMessageId(String session, Direction direction, long se
                 .build();
     }
 
-    protected Timestamp createTimestamp() {
-        Instant now = Instant.now();
+    protected Timestamp toTimestamp(Instant timestamp) {
```
[the remark this answers is not in the document]
Comment (andreydrobynin) [reply src/test/java/com/exactpro/th2/mstore/TestCaseMessageStore.java]: done

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]
Please redirect merge to the dev branch (pre release)

File: src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java [hunk 5]
```diff
@@ -183,15 +187,34 @@ public final void handle(T messageBatch) {
...
+
+            SequenceToTimestamp first = extractSequenceToTimestamp(firstMessage);
+            SequenceToTimestamp last = extractSequenceToTimestamp(lastMessage);
+            SequenceToTimestamp previousBatchLast = sessionData.getAndUpdateLastSequenceToTimestamp(last);
+            if (first.sequenceIsLessOrEquals(previousBatchLast)) {
+                if (logger.isErrorEnabled()) {
+                    logger.error(
+                            "Found batch with less or equal sequence: {}. Last sequence from previous batch: {}",
+                            shortDebugString(messageBatch),
+                            previousBatchLast.getSequence()
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java:205]:
```suggestion
                            previousBatchLast.getSequence(),
                            shortDebugString(messageBatch)

```

File: src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java [hunk 6]
```diff
@@ -183,15 +187,34 @@ public final void handle(T messageBatch) {
...
+                    );
+                }
+                return;
+            }
+            if (first.timestampIsLess(previousBatchLast)) {
+                if (logger.isErrorEnabled()) {
+                    logger.error(
+                            "Found batch with less timestamp: {}. Last timestamp from previous batch: {}",
+                            shortDebugString(messageBatch),
+                            toInstant(previousBatchLast.getTimestamp())
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java:215]:
```suggestion
                            toInstant(previousBatchLast.getTimestamp(),
                            shortDebugString(messageBatch))
```

**Review by Nikita-Smirnov-Exactpro — CHANGES_REQUESTED** [review]

File: src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java (hunk 6 above)
Comment (andrey-drobynin) [reply src/main/java/com/exactpro/th2/mstore/AbstractMessageStore.java]: done

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Pull request merged**